### PR TITLE
TP-link tapo C210 PTZ support

### PR DESCRIPTION
**Pull request by danielvandenberg95**
DO NOT MERGE! (yet?)

I managed to get PTZ  working for tapo C210. It's apparently based on https://github.com/likaci/mercury-ipc-control. I'd like to somehow get it into frigate.

I'm missing "get position" info, so auto-track may not work (or we'll have to make that work without "get position").

I'm kinda seeing two options:
1. Add compatibility for this device as a plugin. How?
2. Add it as a separate type, next to onvif. It doesn't really have anything to do with onvif any more at this point. But quite some code, like handle_command, is common. So, frigate/ptz/common.py, make onvif.py inherit from that, and make mercury-ipc inherit from it too?

(3. not merge it... How big is TP-link/tapo? Upvotes for popularity/importance?)

### CODE REVIEW

**Comment by danielvandenberg95** [comment]
(now off to bed.... It's 04:00 AM here... Whoops...)

**Comment by NickM-27** [comment]
Personally I don't think we're looking to maintain separate ptz implementations for the different proprietary systems out there. Onvif is an open standard and in my / others testing works with the camera brands that are recommended in the frigate docs. I also don't believe any of the maintainers have these cameras so we couldn't maintain this code.

It's possible when we eventually update to a newer version of Python that the onvif-zeep-async dependency will work better with the tapo cameras, as some have reported that home assistant works fine.

**Comment by danielvandenberg95** [comment]
@NickM-27 unfortunately: no. I can confirm that PTZ seems to work through Onvif, but only for a while.
image
This is a capture from while using Onvif device manager. The top red marked area works fine, until suddenly there's a 30 second delay before it responds with a 400. This happens consistently, it really is a bug in the C210. So unless they fix it in a firmware update...

What is your opinion on letting plugins hook into this part of the system?

I have seen that some other camera's have simple http calls for left/right/up/down/stop, and I am seeing that there is already an extension out there, "double-take". I haven't looked into how it should work exactly.

**Comment by NickM-27** [comment]
I'm not worried about if it would be difficult to do, but none of the frigate maintainers have the capacity currently to maintain this plugin or test it in future release or support users who see issues. 

I don't really see it being worth it supporting manufacturer specific http calls for some different camera brands that generally are problematic with frigate anyway. 

Personally, I don't think I agree with the idea in general and would prefer to stick with onvif. It's not up to me though, so we can see what @blakeblackshear thinks.

**Comment by danielvandenberg95** [comment]
Oh, I didn't mean plugins maintained by the frigate team. Just contrib plugins. Well, yeah, could technically speaking treat Onvif as the one and only PTZ plugin maintained by Frigate, that would probably keep any APIs up to date and help with separation of concerns.

While waiting for blakeblackshear, I'll see if I can draft out my ideas a bit more.

**Comment by jmtatsch** [comment]
@danielvandenberg95 Did you reach out to the tapo support already? Maybe fixing this at its source is best. Especially as its older sibling C200 apparently works as expected.

**Comment by danielvandenberg95** [comment]
Good call, forgot the simplest option. I'll do so yeah, see what they have to say.

**Comment by danielvandenberg95** [comment]
As could have been expected from your average customer support:

_​Bedankt voor uw support aanvraag.
De FAQ die er staat geeft aan dat het mogelijk is om via Onvif te werken, echter ondersteunen wij dat niet. 
Ik heb dit toch even doorgezet naar ons hoofdkantoor of hun een idee hebben. Zodra ik een antwoord heb laat ik het weten._

My translation:
_Thank you for your support request.
The FAQ indicates it is possible to use Onvif. However, we do not support this.
I have, however, forwarded this to our main office to see if they have any ideas. As soon as I have an answer I'll let you know._


In my response to this I've added the following two images:
image
image
Even the datasheet says it does. I'll keep this PR up to date... Let's see what they come up with next...

**Comment by jmtatsch** [comment]
I did write the support as well and encourage all other c210 owners here to do the same.
By the way, using the IPCams app with ONVIF stack I get the same 400 error after a few movements.

**Comment by jmtatsch** [comment]
Device got a firmware update to 1.3.7 like two days ago but problem is still not fixed.

**Comment by danielvandenberg95** [comment]
After a whole heap of "not an answer"-answers, I've sent them a notice of default in correspondance to the European Customer Center. They've got 14 days from now... Let's see what they say.

**Comment by jmtatsch** [comment]
I didn't even get an answer from their support 😠

**Comment by danielvandenberg95** [comment]
"We are discussing the issue with the RD Team and we will reply to you this week."

I'm calling that progress...

**Comment by danielvandenberg95** [comment]
Bedankt voor uw geduld. Ons hoofdkantoor reageert:

"We have found the cause of the issue and we are making the firmware to fix it. Please let the customer wait patiently. Thank you."

Met andere woorden het was inderdaad zoals u voor de eerste keer al aangaf een bug in de camera, deze is gevonden en er wordt dus nu gewerkt aan een update die dit probleem moet gaan oplossen voor u.


Google Translated for your convenience:

Thank you for your patience.  Our head office responds:

 "We have found the cause of the issue and we are making the firmware to fix it. Please let the customer wait patiently. Thank you."

 In other words, as you first indicated, it was indeed a bug in the camera, it has been found and we are now working on an update that should solve this problem for you.

**Comment by NickM-27** [comment]
that's encouraging, given this news and the previous discussion above about non onvif implementations I am going to close this PR. feel free to continue commenting if there is news

**Comment by danielvandenberg95** [comment]
V1.3.11 is out according to https://www.tp-link.com/nl/support/download/tapo-c210/#Firmware-Release-Notes but my camera says it's up to date, while updating to max 1.3.9. does anybody already have the ability to test that update? Otherwise I'll just have to wait some more.

**Comment by ehn** [comment]
My C210s aren't seeing an update newer than 1.3.9 either. However, they're hardware version 1, and it seems 1.3.11 is only for hardware version 2.

The changelog for 1.3.11 only includes "Optimized stability of the Wi-Fi connection." Not sure that we should hope for ONVIF fixes in this version.

**Comment by danielvandenberg95** [comment]
Just managed to confirm, 1.3.11 behaviour is different. It now reboots when using onvif device manager. So yeah, let's hope they fixed it in the next one, this would've already been way sooner than I'd have expected from a company like TP-link.

**Comment by ibrahim-mubarak** [comment]
I just setup Tapo C210 HW 2.0, FW 1.3.11 with Frigate 0.13.0 Beta 6
Everything seems to be working smoothly
No reboots or anything
Haven't tested Auto tracking yet, since this is my first Frigate install and I'm just experimenting right now.

**Comment by danielvandenberg95** [comment]
Thats what I meant with "I'm missing "get position" info" I. The original post.

**Comment by Rudd-O** [comment]
It's been a year and a half, multiple firmware updates, multiple releases of Frigate, and ONVIF PTZ still won't work on Frigate with the C210.

Meanwhile Tapo has sold millions of those cameras -- and by their behavior it's clear that they have zero plans to fix ONVIF support in them.

Maybe it's time to resurrect this?  Frigate is a much more important project now, and it would be good to have this camera working properly in Frigate.
